enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,647 @@
+local services, etc. This resource is used to create a discovery image that
+should be used for booting hosts. In the REST API this corresponds to the
+"image" resource that is embedded in the "cluster" resource, but for
+kubernetes-native it is more natural for it to be separate.
+
+The discovery ISO can be downloaded from a URL that is available in the
+resource's Status.
+
+The details of this resource definition were discussed [in a
+pull request](https://github.com/openshift/assisted-service/pull/969).
+
+Summary of fields:
+
+Spec
+* egress proxy settings
+* a list of NTP sources (hostname or IP) to be added to all cluster
+hosts. They are added to any NTP sources that were configured through other
+means.
+* a list of SSH public keys that will be added to all agents for use in debugging.
+* pull secret reference
+* a label selector for Agents. This is how Agent resources are identified as
+belonging to a particular InstallEnv.
+* labels to add to Agent resources upon creation. These labels should match
+the label selector above.
+* Array of ClusterDeployment references. This is a whitelist of ClusterDeployments
+that are authorized to consume Agents from this InstallEnv. See the section below
+on [Namespaces and Multi-tenancy](#namespaces-and-multi-tenancy) for details.
+
+Status
+* a URL to download the discovery ISO produced for this InstallEnv.
+
+**Agent**
+Agent is a new resource, part of the assisted installer's new operator, that
+represents a host that is destined to become part of an OpenShift cluster, and
+is running an agent that is able to run inspection and installation tasks. It
+correlates to the "host" resource in the Assisted Installer's REST API.
+
+The details of this API are available [as
+code](https://github.com/openshift/assisted-service/pull/861), with some
+details yet to be changed.
+
+Summary of fields:
+
+Spec
+* Role that the Node will have
+* Hostname
+* Approved
+* Install device, set automatically when there is platform-integrated
+automation booting the discovery ISO, or otherwise set by the user in a
+boot-it-yourself scenario.
+
+Status
+* Comprehensive inspection and validation data
+
+#### Hive Integration
+
+Hive has a [ClusterDeployment
+CRD](https://github.com/openshift/hive/blob/master/docs/using-hive.md#clusterdeployment)
+resource that represents a cluster to be created. It includes a "platform"
+section where platform-specific details can be captured. For an agent-based
+workflow, this section will include a new `AgentBareMetal` platform that
+contains such fields as:
+
+* API VIP
+* API VIP DNS name
+* IngressVIP
+
+A new `InstallStrategy` section of the Spec enables the API to describe a way
+of installing a cluster other than the default of using `openshift-install`.
+The new field has an "agent" option that would be common to any Agent-driven
+installation, regardless of platform. Fields include:
+
+* AgentSelector, a label selector for identifying which Agent resources should
```

Comment:
Same as above @dgoodwin @filanov

Reply:
This one is a little different, for late binding we're going to be looking to fulfill machine requests from a pool of available agents. That pool might contain different hardware perhaps or geolocated systems? In which case we'd want to say choose from Agents with sub-sets of labels and indeed mutliple cluster could be competing for those resources.

Once assigned to a cluster though, we will need an explicit  reference from Agent -> ClusterDeployment.

Is both a reasonable approach here?

Reply:
I agree that many-to-many is the intent here. This could be the use case for having a ClusterDeployment reference on an Agent so that it can be formally claimed.

Reply:
For now a ClusterDeployment will correspond to exactly one InstallEnv. Once we have late binding we can add a ClusterDeployment reference onto the Agent for use in claiming it.

Reply:
We'd have to have some way of handling "legacy" claimed Agents with no ClusterDeployment reference when we add late-binding. Since we know claiming will be needed, better to add the reference now?

Reply:
Agents will be garbage-collected after being added to clusters, so that should take care of any need to handle legacy Agents.

Reply:
> Agents will be garbage-collected after being added to clusters, so that should take care of any need to handle legacy Agents.

Except any unclaimed agents that exist from before an upgrade.

Reply:
We know from both the Agent state on its CR and from the assisted service DB whether an agent has started an installation process or not, and whether we're done interacting with it. After an upgrade to a future where we've added the ClusterDeployment reference to the Agent CRD, it should be straight-forward for the controller to, upon reconciling each Agent: delete any lingering ones that are complete, or add the correct ClusterDeployment resource if by some chance an agent was in the middle of installing when the service got upgraded.

If there's time in the immediate release cycle, it might still make more sense to add the ClusterDeployment reference in the first release so we can avoid worrying about migration. But I also think it's safe to defer if we need to.

Reply:
Ok, let's move on